core/src/test/java/org/apache/iceberg/util/TestFileSystemWalker.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.types.Types;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+public class TestFileSystemWalker {
+
+  @TempDir private Path tempDir;
+  private String basePath;
+  private Configuration hadoopConf;
+  private Map<Integer, PartitionSpec> specs;
+  private HadoopFileIO fileIO;
+
+  @BeforeEach
+  public void before() throws IOException {
+    this.basePath = tempDir.toAbsolutePath().toString();
+    this.hadoopConf = new Configuration();
+    this.fileIO = new HadoopFileIO(hadoopConf);
+
+    Schema schema =
+        new Schema(
+            Types.NestedField.required(1, "id", Types.IntegerType.get()),
+            Types.NestedField.required(2, "data", Types.StringType.get()));
+    PartitionSpec spec = PartitionSpec.builderFor(schema).identity("id").build();
+    specs = ImmutableMap.of(0, spec);
+
+    Files.createDirectories(Paths.get(basePath, "normal_dir"));
+    Files.createDirectories(Paths.get(basePath, "normal_dir/dep1"));
+    Files.createDirectories(Paths.get(basePath, "hidden_dir/_partition"));
+    Files.createFile(Paths.get(basePath, "file1.txt"));
+    Files.createFile(Paths.get(basePath, "normal_dir/file2.txt"));
+    Files.createFile(Paths.get(basePath, "normal_dir/dep1/file3.txt"));
+    Files.createFile(Paths.get(basePath, "hidden_dir/.hidden_file.txt"));
+    Files.createFile(Paths.get(basePath, "hidden_dir/_partition/file3.txt"));
+  }
+
+  @Test
+  public void testListDirRecursivelyWithHadoop() {
+    List<String> foundFiles = new ArrayList<>();
+    List<String> remainingDirs = new ArrayList<>();
+    Predicate<FileStatus> fileFilter =
+        fileStatus -> fileStatus.getPath().getName().endsWith(".txt");
+    FileSystemWalker.listDirRecursivelyWithHadoop(
+        basePath,
+        specs,
+        fileFilter,
+        hadoopConf,
+        Integer.MAX_VALUE,
+        Integer.MAX_VALUE,
+        remainingDirs::add,
+        foundFiles::add);
+
+    assertThat(foundFiles).hasSize(3);
+    assertThat(foundFiles).contains(Paths.get("file://", basePath, "file1.txt").toString());
+    assertThat(foundFiles)
+        .contains(Paths.get("file://", basePath, "normal_dir/file2.txt").toString());
+    assertThat(foundFiles)
+        .contains(Paths.get("file://", basePath, "normal_dir/dep1/file3.txt").toString());
+    assertThat(remainingDirs).isEmpty();
+  }
+
+  @Test
+  public void testListDirRecursivelyWithHadoopMaxDepth() {
+    List<String> foundFiles = new ArrayList<>();
+    List<String> remainingDirs = new ArrayList<>();
+    Predicate<FileStatus> fileFilter =
+        fileStatus -> fileStatus.getPath().getName().endsWith(".txt");
+    FileSystemWalker.listDirRecursivelyWithHadoop(
+        basePath,
+        specs,
+        fileFilter,
+        hadoopConf,
+        2, // maxDepth
+        10, // maxDirectSubDirs
```

Comment:
Could we add tests for this?

Reply:
I see, that we have test for `maxDepth`, but we still need test for `maxDirectSubDirs`